aws/src/main/java/org/apache/iceberg/aws/HttpClientCache.java

```diff
@@ -0,0 +1,212 @@
+ * under the License.
+ */
+package org.apache.iceberg.aws;
+
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.Supplier;
+import org.apache.iceberg.relocated.com.google.common.annotations.VisibleForTesting;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import software.amazon.awssdk.http.ExecutableHttpRequest;
+import software.amazon.awssdk.http.HttpExecuteRequest;
+import software.amazon.awssdk.http.SdkHttpClient;
+
+/**
+ * A cache that manages the lifecycle of shared HTTP clients for AWS SDK v2 using reference
+ * counting. Package-private - only accessed via {@link BaseHttpClientConfigurations}.
+ */
+final class HttpClientCache {
+  private static final Logger LOG = LoggerFactory.getLogger(HttpClientCache.class);
+
+  private final ConcurrentMap<String, ManagedHttpClient> clientMap;
+
+  private static volatile HttpClientCache instance;
+
+  static HttpClientCache getInstance() {
+    if (instance == null) {
+      synchronized (HttpClientCache.class) {
+        if (instance == null) {
+          instance = new HttpClientCache();
+        }
+      }
+    }
+    return instance;
+  }
+
+  private HttpClientCache() {
+    this.clientMap = Maps.newConcurrentMap();
+  }
+
+  /**
+   * Get or create a managed HTTP client for the given configuration. Each call increments the
+   * reference count for the client and returns a ref-counted wrapper.
+   *
+   * @param clientKey unique key identifying the client configuration
+   * @param clientFactory factory to create the HTTP client if not cached
+   * @return a ref-counted HTTP client wrapper
+   */
+  SdkHttpClient getOrCreateClient(String clientKey, Supplier<SdkHttpClient> clientFactory) {
+    ManagedHttpClient managedClient =
+        clientMap.computeIfAbsent(
+            clientKey,
+            k -> {
+              LOG.debug("Creating new managed HTTP client for key: {}", k);
+              SdkHttpClient httpClient = clientFactory.get();
+              return new ManagedHttpClient(httpClient, k);
+            });
+    // Return the cached ref-counted wrapper
+    return managedClient.acquire();
+  }
+
+  /**
+   * Release a reference to the HTTP client. When the reference count reaches zero, the client is
+   * closed and removed from the cache.
+   *
+   * @param clientKey the key identifying the client to release
+   */
+  void releaseClient(String clientKey) {
+    ManagedHttpClient managedClient = clientMap.get(clientKey);
+    if (managedClient != null) {
+      if (managedClient.release()) {
+        // Client was closed, remove from map
+        clientMap.remove(clientKey, managedClient);
+      }
+    }
+  }
+
+  @VisibleForTesting
+  ConcurrentMap<String, ManagedHttpClient> clientMap() {
+    return clientMap;
+  }
+
+  @VisibleForTesting
+  void shutdown() {
```

Comment:
A bit of a nit here, but this doesn't feel like a shutdown, but rather a `clear()` method.  It's still usable after shutdown, so the behavior isn't quite what I would expect.